### PR TITLE
C++ benchmarking

### PR DESCRIPTION
**Pull request by novikov-alexander**
- Documentation including methodology and architecture of benchmark
- Shared libraries (modules) wrapping Manual GMM, Manual Eigen GMM, Manual Eigen BA and Manual Eigen Hand objectives according to documentation.
- Non-Eigen version of Manual Ba and Manual Hand objectives.
- Manual LSTM module.
- C++ runner which benchmarks functions from modules according to documentation.
- Global runner improvements to support c++ runner.
- Module and runner unit tests using GTest (includes tests of module loading, calculation correctness, tests of module main loop optimization and tests of measurements correctness)
- Linux compatibility

### CODE REVIEW

File: docs/Architecture.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+On the highest level ADBench consists of
+- Global runner
+- Benchmark runners
+- Automatic differentiation (AD) framework testing modules
+- Result-processing scripts
+
+_Testing modules_ here are modules in terms of the platform they are developed for (e.g shared objects for C++, assemblies for .NET (Core), `.py` files in Python) that perform AD using the frameworks being tested.
+
+_Benchmark runners_ are console applications that load _testing modules_ and input data and measure the time modules take to compute the objective function and its considered derivative for the loaded input data. Then they write measured times and computed derivatives to files with standardized names. We have one benchmark runner per development platform, so that we can use the same time measuring code for all frameworks supporting that platform.
+
```
Comment (awf) [inline docs/Architecture.md]: time-measuring

File: docs/Architecture.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+It contains code for starting all the _benchmark runners_ and is aware of all existing _testing modules_.
+
+_Global runner_ is started by the user. User may pass some or all of the variables listed in the [methodology](Methodology.md) as command line arguments.
+
+__The exact specifications are to be developed along with the _global runner_ itself__.
+
+### Result-Processing Scripts
+
+Two scripts. One looks at the ouput files and compares the results in them to the manually computed correct ones, the other produces visualizations of the timeings.
+
```
Comment (awf) [inline docs/Architecture.md]: timings

File: docs/Methodology.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Benchmarking Methodology
+
+This document describes the process of obtaining time measurements for computations of objective functions and their derivatives using different automatic differentiation (AD) frameworks.
+
+Benchmarking process depends on the following variables:
+- `nruns_F` - maximum number of times to run the computation the objective function for timing
+- `nruns_J` - maximum number of times to run the computation of the considered derivative (gradient or Jacobian) for timing
+- `time_limit` - _soft_ (see below) time limit for benchmarking the computation of either the objective function or its gradient/Jacobian
+- `timeout` - _hard_ time limit for the complete (including time that is not measured) benchmark
+- `minimum_measurable_time` - minimum time, computation needs to run to produce a reliable measurement
```
Comment (awf) [inline docs/Methodology.md]: s/,/that/

**Comment by iliaeg** [comment]
> partial comments. Are graphs working? If so, it would be great to paste one in to the PR.

As far as I know they should be working. So now I produce data to plot them, it's quite long process.

**Comment by awf** [comment]
Great, thanks!
Here's the GMM Jacobian 10K plot, which looks good (i.e. as expected).
image
And 1K...
image
And hand...
image

**Comment by msdmkats** [comment]
I've a question. The last commit here adds a couple MB of pngs with plots. @awf , do we actually want them in this repository? There was an issue #11 where it was decided not to store them.

**Comment by tomjaguarpaw** [comment]
Indeed. Please do not store plots (or any form of build artifacts) in the repo.

**Comment by msdmkats** [comment]
One more question, before I merge this. Merge commit, rebase, or squash?
Github shows conflicts, because at some early point in branch's history master was merged into it, instead of rebasing it on top of master. I fixed it on my machine by doing a proper rebase, so rebase and merge is a possibility.

**Comment by tomjaguarpaw** [comment]
Rebase and merge sounds ideal.  I guess you will have to force push the rebased branch from your machine because GitHub doesn't seem to be able to do it itself.

**Pull request merged**